src/app/cli/src/commands/verify_command.rs

```diff
@@ -79,17 +82,35 @@ impl VerifyCommand {
     ) -> GenericVerificationResult {
         let dataset_ref_pattern = self.refs.first().unwrap();
 
-        let requests: Vec<_> = filter_datasets_by_pattern(
+        let dataset_ids: Vec<_> = filter_datasets_by_pattern(
             self.dataset_repo.as_ref(),
             vec![dataset_ref_pattern.clone()],
         )
-        .map_ok(|dataset_handle| VerificationRequest {
-            dataset_ref: dataset_handle.as_local_ref(),
-            block_range: (None, None),
-        })
+        .map_ok(|dataset_handle| dataset_handle.as_local_ref().id().unwrap().clone())
         .try_collect()
         .await?;
 
+        let requests: Vec<_> = if self.recursive {
+            self.dependency_graph_service
+                .get_all_upstream_dependencies(dataset_ids)
+                .await
+                .int_err()?
+                .map(|dataset_id| VerificationRequest {
+                    dataset_ref: DatasetRef::ID(dataset_id),
+                    block_range: (None, None),
+                })
+                .collect()
+                .await
```

Comment:
I wonder if we care about the order of verification here.
Right now the order is undetermined, since goes through a hash set.
But it seems to me that the order is important, and we need to verify the most upstream datasets first and fail early, if there is any problem with those - the datasets from next levels are unlikely good, if 1st level is wrong.

For "delete" command, the order definitely matters. Each deletion will check if any downstream datasets are still dangling, and will fail.